compiler/rustc_metadata/src/locator.rs

```diff
@@ -882,8 +882,28 @@ crate enum CrateError {
     NonDylibPlugin(Symbol),
 }
 
+// Used to work around a borrowck issue where the session can't be borrowed at the same time as the loader.
+crate enum LoaderOrSess<'a, 'b> {
+    Loader(&'b mut CrateLoader<'a>),
+    Session(&'b Session),
+}
```

Comment:
This looks pretty bad.
Could the resolution of `core` be moved to the call site of `err.report()` instead? (So `missing_core: bool` will be passed as a flag to `report`.)

Reply:
Agreed, I'm not super happy with it either.

>  Could the resolution of core be moved to the call site of err.report() instead? (So missing_core: bool will be passed as a flag to report.)

That means it will try loading `core` unconditionally on crate errors - is that ok?

Reply:
That's ok.

Of course, you can optimize and load it only in the case of `matches!(err, CrateError::LocatorCombined(..)) && crate_name != sym::core`, but it shouldn't be necessary because it only happens during erroneous compilation and `core` usually is already loaded and in the cache anyway.

Reply:
Ok, the logic is a lot simpler now :)